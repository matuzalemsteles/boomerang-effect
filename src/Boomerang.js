/**
 * Copyright (c) 2018, Matuzal√©m Teles.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Config} from 'metal-state';
import Component from 'metal-jsx';

let Boomerang = {};

class Provider extends Component {
	static PROPS = {
		events: Config.object(),
	};

	getChildContext() {
		return {
			boomerang: {
				events: this.props.events,
			},
		};
	}

	render() {
		const {children} = this.props;

		return children;
	}
}

Boomerang.Provider = Provider;

export {Provider};

export default Boomerang;
