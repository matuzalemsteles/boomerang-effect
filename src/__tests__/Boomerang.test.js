/**
 * Copyright (c) 2018, MatuzalÃ©m Teles.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Config} from 'metal-state';
import Boomerang, {connect} from '../Boomerang';
import Component from 'metal-jsx';

describe('Boomerang', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should call the event from context and update the Provider state', () => {
		jest.useFakeTimers();
		class Child extends Component {
			_handleClickButton() {
				this.emit('buttonClicked');
			}

			render() {
				return (
					<button onClick={this._handleClickButton.bind(this)}>
						Click me!
					</button>
				);
			}
		}

		const ChildWithConnect = connect(Child);

		class Provider extends Component {
			static STATE = {
				clicked: Config.number().value(0),
			};

			_handleClickButton() {
				this.state.clicked++;
			}

			render() {
				const {children} = this.props;
				const events = {
					buttonClicked: this._handleClickButton.bind(this),
				};

				return (
					<Boomerang.Provider events={events}>
						{children}
					</Boomerang.Provider>
				);
			}
		}

		class ChildWithChild extends Component {
			render() {
				return <ChildWithConnect />;
			}
		}

		class App extends Component {
			render() {
				return (
					<Provider ref="provider">
						<ChildWithChild />
					</Provider>
				);
			}
		}

		component = new App({});
		component.element.click();
		jest.runAllTimers();

		const {provider} = component.refs;

		expect(provider.state.clicked).toBe(1);
	});

	it('should pass the context events to the component', () => {
		class Child extends Component {
			render() {
				return <p>Foo</p>;
			}
		}

		const ChildWithConnect = connect(Child);

		class App extends Component {
			render() {
				const events = {
					clicked: () => {},
				};

				return (
					<Boomerang.Provider events={events}>
						<ChildWithConnect ref="child" />
					</Boomerang.Provider>
				);
			}
		}

		component = new App({});

		expect(component.refs.child.refs.connect.props.events).toMatchObject({
			clicked: expect.any(Function),
		});
	});

	it('should pass the context events together with those of the props for the component', () => {
		class Child extends Component {
			render() {
				return <p>Foo</p>;
			}
		}

		const ChildWithConnect = connect(Child);

		class App extends Component {
			render() {
				const events = {
					clicked: () => {},
				};

				const Events = {
					close: () => {},
				};

				return (
					<Boomerang.Provider events={events}>
						<ChildWithConnect events={Events} ref="child" />
					</Boomerang.Provider>
				);
			}
		}

		component = new App({});

		expect(component.refs.child.refs.connect.props.events).toMatchObject({
			clicked: expect.any(Function),
			close: expect.any(Function),
		});
	});

	it('should must pass the props to component', () => {
		class Child extends Component {
			render() {
				return <p>Foo</p>;
			}
		}

		const ChildWithConnect = connect(Child);

		class App extends Component {
			render() {
				const events = {
					clicked: () => {},
				};

				return (
					<Boomerang.Provider events={events}>
						<ChildWithConnect foo="bar" ref="child" />
					</Boomerang.Provider>
				);
			}
		}

		component = new App({});

		expect(component.refs.child.refs.connect.props.foo).toBe('bar');
	});

	it('should must pass the children`s to component', () => {
		class Child extends Component {
			render() {
				return <p>{this.props.children}</p>;
			}
		}

		const ChildWithConnect = connect(Child);

		class App extends Component {
			render() {
				const events = {
					clicked: () => {},
				};

				return (
					<Boomerang.Provider events={events}>
						<ChildWithConnect foo="bar" ref="child">
							<h1>Hello World!</h1>
						</ChildWithConnect>
					</Boomerang.Provider>
				);
			}
		}

		component = new App({});

		expect(component).toMatchSnapshot();
	});
});
