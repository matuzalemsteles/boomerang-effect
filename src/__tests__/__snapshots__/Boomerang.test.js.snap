// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boomerang should must pass the children\`s to component 1`] = `
<p>
  <h1>Hello World!</h1>
</p>
`;
